### PR TITLE
#184347036 add item to cart

### PR DESCRIPTION
**Pull request by irakozetony**
### What does this PR do?
Users should be able to add items to the cart

### Description of Task to be completed?
Check if a user is a buyer or admin and allow them to add products to the cart, and calculate the total

### How should this be manually tested?
1. Checkout to `feature/add-to-cart-184347036`
2. Run `npm run seeding`
3. Run `npm run dev`
4. Go to the ` http://localhost:<port>/api/v1/cart/add-to-cart` URL in Postman
5. Try adding a product to the cart

#### What are the relevant pivotal tracker stories?
[#184347036](https://www.pivotaltracker.com/story/show/184347036)

### CODE REVIEW

**Review by fabmukunzi — APPROVED** [review]

**Review by Abdden — APPROVED** [review]

**Review by nshimiyejayd — APPROVED** [review]

File: src/controllers/cart.controller.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+export class CartController {
+  static async addToCart(req, res) {
+    try {
+      const { cartProducts, cartTotal } = req;
```
Comment (TuyizeeAnastase) [inline src/controllers/cart.controller.js]: where are you getting this data from body, query, and params?

File: src/controllers/cart.controller.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+      const productData = cartProducts.map(product => ({
+        name: product.dataValues.name,
+        price: product.dataValues.price,
+				quantity: product.quantity,
+        images: product.dataValues.images,
+      }));
+
```
Comment (TuyizeeAnastase) [inline src/controllers/cart.controller.js]: no need to use dataValues

File: src/controllers/cart.controller.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      return res.status(200).json({
```
Comment (TuyizeeAnastase) [inline src/controllers/cart.controller.js]: status code is not 200

File: src/middlewares/cart.middleware.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        existingProduct.quantity += 1;
+      } else {
+        products.push({ quantity: 1, productId: productId });
+      }
+      await CartService.updateCart({ products }, userCart.id);
+    }
+    req.cart = userCart;
+    return next();
+  } catch (error) {
+    console.error(error);
```
Comment (TuyizeeAnastase) [inline src/middlewares/cart.middleware.js]: Don't push console

File: src/middlewares/cart.middleware.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        products,
+      };
+      userCart = await CartService.createCart(cart);
+    } else {
+      let products = userCart.products;
+      let existingProduct = products.find(p => p.productId === productId);
+      if (existingProduct) {
+        existingProduct.quantity += 1;
+      } else {
+        products.push({ quantity: 1, productId: productId });
```
Comment (TuyizeeAnastase) [inline src/middlewares/cart.middleware.js]: you need to check the condition of the added product when quantity>1

File: src/middlewares/product.middleware.js [hunk 6]
```diff
@@ -11,6 +11,15 @@ export const checkIfProductExists = async (req, res, next) => {
   else next();
 };
 
+export const checkProductAvailable = async(req, res, next)=>{
+  const { productId } = req.body;
+  const getProduct = await Products.findOne({
+    where: {id: productId},
+  });
+  if(getProduct) next();
+  else return res.status(404).json({message: 'Product not found'});
```
Comment (TuyizeeAnastase) [inline src/middlewares/product.middleware.js]: first start with condition not found else call next()

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

File: src/middlewares/cart.middleware.js (hunk 5 above)
Comment (irakozetony) [reply src/middlewares/cart.middleware.js]: Here,
You mean, like when user adds quantity?
Currently what am doing is, if a user adds a product that's already in the cart, I increment quantity by one

**Review by TuyizeeAnastase — APPROVED** [review]

File: src/middlewares/cart.middleware.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		const cart = {
+			userId,
+			total,
+			products,
+		};
+		userCart = await CartService.createCart(cart);
+	}
+	req.cart = userCart;
+	next();
+};
```
Comment (ncutixavier) [inline src/middlewares/cart.middleware.js]: This middleware should only be used to get the user's cart. 
Hint: And if there is no cart, return next()

File: src/controllers/cart.controller.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        price: product.price,
+				quantity: product.quantity,
+        images: product.images,
+      }));
+
+      return res.status(201).json({
+        products: productData,
+        cartTotal,
+        message: 'Item added to cart successfully',
+      });
```
Comment (ncutixavier) [inline src/controllers/cart.controller.js]: - This controller should hold the functionality to create a cart

File: src/controllers/cart.controller.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    } catch (error) {
+      return res.status(500).json({
+        error: error.message,
+        message: 'Something went wrong, try again',
```
Comment (ncutixavier) [inline src/controllers/cart.controller.js]: Be specific, what's wrong?

**Review by ncutixavier — CHANGES_REQUESTED** [review]

**Review by ncutixavier — APPROVED** [review]
LGTM

**Pull request merged**